### PR TITLE
fix issue 16597 - deprecate `shared` as associative array storage class

### PR DESCRIPTION
**Pull request by ghost**
It's not possible to output directly an error because in theory it would be possible that someone uses a shared AA in combination with synchronization barriers, so instead encourage using  `__gshared`.

### CODE REVIEW

File: changelog/osx-32bit.dd [hunk 1]
```diff
@@ -1,9 +1,7 @@
-OSX 32-bit is no longer supported
+`shared` associate arrays are deprecated
```
Comment (Basile-z) [inline changelog/osx-32bit.dd]: wow, such a successful copy and paste

**Comment by jacob-carlborg** [comment]
Is this in the spec?

**Comment by Basile-z** [comment]
Not to my knowledge. To be perfectly honest, this PR is naively motivated by a comment in the issue.
The idea is just to avoid head scratching for the people trying to do something that cant work.

File: src/dmd/dsymbolsem.d [hunk 2]
```diff
@@ -694,6 +694,11 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                 dsym.error("no definition of struct `%s`", ts.toChars());
             }
         }
+        if (dsym.type.isTypeAArray && (dsym.type.isShared || dsym.storage_class & STC.shared_))
+        {
+            dsym.deprecation("associative arrays are not thread safe and cant be declared `shared`." ~
```
Comment (ZombineDev) [inline src/dmd/dsymbolsem.d]:
```suggestion
            dsym.deprecation("associative arrays are not thread safe and can't be declared `shared`." ~
```

File: test/fail_compilation/deprecate16597.d [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// REQUIRED_ARGS: -de
+/* TEST_OUTPUT:
+---
+fail_compilation/deprecate16597.d(9): Deprecation: variable `deprecate16597.aa1` associative arrays are not thread safe and cant be declared `shared`.Use `__gshared` and barriers instead
+fail_compilation/deprecate16597.d(10): Deprecation: variable `deprecate16597.aa2` associative arrays are not thread safe and cant be declared `shared`.Use `__gshared` and barriers instead
+---
+*/
+// https://issues.dlang.org/show_bug.cgi?id=16597
+shared int[int] aa1;
+shared(int[int]) aa2;
```
Comment (ZombineDev) [inline test/fail_compilation/deprecate16597.d]: We should probably also disallow `shared(V)[K]`, as I don't think the AA implementation takes any precautions to guard reads/writes of shared values. Can you add a test for this as well?

**Review by PetarKirov — APPROVED** [review]

**Comment by Basile-z** [comment]
Sorry i don't feel it anymore. I keep the branch if some one wants to polish and is more convinced about the validity of this restriction.

**Pull request closed without merging**